muikku-rest/src/main/java/fi/otavanopisto/muikku/rest/user/UserRESTService.java

```diff
@@ -268,7 +269,20 @@ public Response getUserEntityProperties(@PathParam("USERENTITYID") Long userEnti
   @RESTPermit (handling = Handling.INLINE, requireLoggedIn = true)
   public Response setUserEntityProperty(fi.otavanopisto.muikku.rest.model.UserEntityProperty payload) {
     UserEntity loggedUserEntity = sessionController.getLoggedUserEntity();
-    userEntityController.setUserEntityProperty(loggedUserEntity, payload.getKey(), payload.getValue());
+    UserEntity userEntity = userEntityController.findUserEntityById(payload.getUserEntityId());
+    Boolean isStudent = userEntityController.isStudent(userEntity);
+    Boolean isLoggedUserStudent = userEntityController.isStudent(loggedUserEntity);
+
+    if (payload.getUserEntityId() != null && payload.getKey().equals("hopsPhase")) { // This is for hops.phase. Only staff members can set phases.
+      if (isStudent && !isLoggedUserStudent) { // If userEntity from payload is student & logged user is not student
+        userEntityController.setUserEntityProperty(userEntity, payload.getKey(), payload.getValue());
+      }
+      
+    } else if (payload.getUserEntityId() != null) { // This is for student graduation date goal
+      userEntityController.setUserEntityProperty(userEntity, payload.getKey(), payload.getValue());
```

Comment:
If the currently logged in user is a student, the user entity id in the payload must match the logged user entity id.